### ISSUE TITLE
OPS - Create Staging and Production branches for deployment 

### ISSUE THREAD

**Issue by traycn** [hackforla/311-data#1867]
### Overview

We need to create a Staging and Production branch to establish a deployment pipeline to test new changes before deploying to Production.

### Action Items

- [x] Review "Proposed Instructions for Cherry-Picking from 311-data to 311-landingpage"
- [x] Review "Reddit: How do I create a staging site for github pages"
- [x] Review Github Environments, determine if any of these features can be utilized for this purpose
- [x] Write up a document in the shared drive on which strategy is preferred
  - [x] include explanation on how the chosen strategy retains commit histories from the repo
- [ ] get process approval from Bonnie
- [ ] implement deployment strategy and create staging, production deployments

### Resources/Instructions

- [Example branch of `rebase-merge` of Website team](https://github.com/JessicaLucindaCheng/data-science/commits/rebase-merge)
- https://docs.github.com/en/code-security/getting-started/adding-a-security-policy-to-your-repository
- https://github.com/hackforla/website/issues/6595
- https://docs.github.com/en/pages/getting-started-with-github-pages/configuring-a-publishing-source-for-your-github-pages-site
- https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-deployments/managing-environments-for-deployment

<details><summary>Proposed Instructions for Cherry-Picking</summary>
<p>

Staging Branch: - `/311-data/dev`
Production Branch: - `/311-data/main`

- [ ] Setup branch security as necessary for `311-data/main`
- [ ] Review Git history to prepare for commit history transfer
- [ ] Setup branch security as necessary for `/311-data/dev`
- [ ] Setup automatic transfer of commit history changes from Staging to Production. 

</p>
</details> 

<details><summary>Reddit: How do I create a staging site for github pages</summary>
<p>

Source: https://www.reddit.com/r/github/comments/1c2aly0/how_do_i_create_a_staging_site_for_github_pages/

---

GitHub doesn't natively support this as far as I know (or I haven't figured out how to do this).

So you either need to use third party hosting like Netlify or Vercel, or you hack something together yourself.

I did the latter. I build my projects with Vite, meaning I can specify a base path for my app in the build process. So what I did was just to add staging/ to the already configured base using Regex in the build pipeline.

Then I bought a domain and didn't need the base anymore for the production deployment. My solution to that was to use environment variables. You can access process.env during the build process, so I just put a base path into a .env file when building for staging, or I don't do that when building for production. For staging, I then end up with /staging/ as the base, and for production, it stays as undefined, meaning /. This works perfectly for me.

Here's the workflow, if you want to copy it: https://github.com/Eisvana/EisvanaWikiPageCreator/blob/main/.github/workflows/build-deploy.yml

The code for Vite is in the Vite config file at the root of the project.

So when I push something to my staging branch, it ends up at domain.com/staging/, and when I push to production, it gets deployed to domain.com/.

</p>
</details>

**Comment by ryanfchase** [comment]
We'll need to review this guide written by website team: https://github.com/hackforla/website/issues/6595

**Comment by ryanfchase** [comment]
Bonnie is going to give us the green light on modifying the 311-landingpage repo, or creating new repos, to use as our "prod site". But she wants a written explanation of how it's going to work. We'll need to explain our strategy for incorporating commits from staging to prod WHILE RETAINING COMMIT AUTHORS. The website team has provided a guide on how to achieve this: https://github.com/hackforla/website/issues/6595#issuecomment-2047453668

Requesting review from dev lead.

**Comment by traycn** [comment]
Great to know. To clarify, this issues now involves the following work:

- establishing 311-landingpage repo as PRODUCTION
- establishing 311-data repo as STAGING
- Transfer commit history from 311-data/main > 311-landingpage/main

Post commit history transfer:
- cherry-picking commits from 311-data/main > 311-landingpage/main

Documentation:
- Documentation on process from How To establish pipeline and the CI/CD process.

**Comment by traycn** [comment]
I have edited ticket details with the definition of the `Local` `Staging` `Production` branches. I've also updated the Action Items to establish our environments while incorporating HackForLA org requirements.

**Comment by cottonchristopher** [comment]
Hi @traycn ,

Please leave a comment with the following items when you have a chance:

- updated ETA
- progress from the last week (if applicable)
- availability for communications during the week

Thanks

**Comment by ExperimentsInHonesty** [comment]
@traycn 

Hi Tracy, this is the template I want teams to use for their weekly updates.  You can always add more info to the items below 1-4, which are the required minimum.

<details><summary>Instructions</summary>

1. Progress: "What is the current status of your project? What have you completed and what is left to do?"
2. Blockers: "Difficulties or errors encountered."
3. Availability: "How much time will you have this week to work on this issue?"
4. ETA: "When do you expect this issue to be completed?"
5. Pictures (if necessary): "Add any pictures that will help illustrate what you are working on."
</details> 

1. Progress: 
2. Blockers: 
3. Availability:
4. ETA:
5. Pictures (if necessary):

**Comment by ryanfchase** [comment]
Questions on Action Items

> - [ ] Review GIT history to prepare for commit history transfer

Anything we are looking for in particular? What should we be looking for?

> - [ ] Check/Setup branch security as necessary for 311-landingpage/main (Production)

What security are we looking for? Any keywords to look into?

> - [ ] Automatically transfer further commit history changes from Staging > Production.

Does the comment specify a way to do this automatically?

**Comment by ExperimentsInHonesty** [comment]
- Review this guide to creating community pages https://100automations.org/guides/community-support-for-automations.html

**Comment by ExperimentsInHonesty** [comment]
Recorded to the cloud Tracy's KT comments on 2025-02-15 11:50-:12:00

**Comment by ryanfchase** [comment]
- adding this document to review, concerning branch security: https://github.com/hackforla/311-data/security/policy
  - which points to: https://github.com/hackforla/website/blob/gh-pages/.github/pull_request_template.md

**Comment by DrAcula27** [comment]
ETA: August 27, 2025
Availability: weekdays after 4pm Pacific time

**Comment by DrAcula27** [comment]
1. Progress: reviewed "Proposed Instructions for Cherry-Picking from 311-data to 311-landingpage" and "Reddit: How do I create a staging site for github pages" posted in the issue.
2. Blockers: N/A
3. Availability: weekdays after 3pm Pacific time
4. ETA: still 8/27

Initial ideas/questions:
- I know this ticket is a bit older, are we still wanting to use this naming convention?
  - Staging Branch: - `/311-data/main`
  - Production Branch: - `HackForLA/311-landingpage/main`
- Do we want to add a `dev` branch before staging as well?
- Under `Proposed Instructions for Cherry-Picking from 311-data to 311-landingpage` in the issue, are these two items not the same task?
  - Setup branch security as necessary for 311-data/main
  - Setup branch security as necessary for hackforla/311-data and 311-landingpage/main

**Comment by DrAcula27** [comment]
Per discussion in the meeting on `20-August-2025`, 
- create a branch called `dev` for staging
- use `main` as production
- go through all resources listed in issue, deep dive on security measures, and have understanding ready to present next week.
- update tasks to remove redundant task

**Comment by DrAcula27** [comment]
# GitHub Security Measures Summary

## Overview

GitHub provides a comprehensive security framework designed to protect repositories, facilitate responsible vulnerability disclosure, and maintain healthy open-source communities. The security measures span four main areas:

1. **Repository Security Configuration** - Technical controls for code protection
2. **Vulnerability Disclosure & Management** - Processes for handling security issues
3. **Community Health & Governance** - Standards and policies for safe collaboration
4. **Automated Security Monitoring** - Tools for continuous security assessment

## Core Security Components

### 1. Security Policies (SECURITY.md)

**Purpose**: Provide clear instructions for reporting security vulnerabilities

**Implementation Steps**:
- Navigate to repository → Security tab → Policy → Start setup
- Create `SECURITY.md` file in repository root, `docs`, or `.github` folder
- Include information about:
  - Supported versions of your project
  - How to report vulnerabilities
  - Contact information for security team
  - Expected response timeline
- To make it easy to find for users, link to security policy from `README.md` and/or other repository docs as needed

**Key Benefits**:
- Establishes clear vulnerability reporting channels
- Reduces confusion for security researchers
- Creates standardized communication process

### 2. Repository Security Advisories

**Purpose**: Enable private collaboration on vulnerability fixes before public disclosure

**Implementation Steps**:
- Create draft security advisory for private discussion
- Collaborate on fixes using temporary private forks
- Request CVE identification numbers (GitHub is a CVE Numbering Authority)
- Publish advisory once patch is ready
- Advisory automatically feeds into GitHub Advisory Database

**Key Features**:
- Private vulnerability discussion
- CVE assignment within 72 hours
- Integration with Dependabot alerts
- Credit attribution for contributors
- REST API support for automation

### 3. Coordinated Vulnerability Disclosure (CVD)

**Core Principle**: Security vulnerabilities should be handled privately until fixes are available

**Process Flow**:
1. Security researcher discovers vulnerability
2. Private report submitted to maintainers
3. Maintainer validates and develops fix
4. Collaborative testing and validation
5. Public disclosure with patch release
6. Community notification and updates

**GitHub Tools Supporting CVD**:
- Private Vulnerability Reporting (PVR)
- Draft security advisories
- Temporary private forks for collaboration
- Automated Dependabot alerts for downstream users

### 4. Repository Security Configuration

**Access Management**:
- Set repository visibility (public/private)
- Configure team and user permissions
- Enable branch protection rules
- Manage access levels (read, write, admin)

**Dependency Security**:
- **Dependency Graph**: Enable to track project dependencies
- **Dependabot Alerts**: Automated vulnerability notifications
- **Dependabot Security Updates**: Automatic pull requests for security patches
- **Dependabot Version Updates**: Keep dependencies current
- **Dependency Review**: Visualize dependency changes in pull requests

**Code Security Features**:
- **Code Scanning**: Automated vulnerability detection using CodeQL
- **Default Setup**: GitHub automatically configures scanning
- **Advanced Setup**: Custom workflow configuration
- **Copilot Autofix**: AI-powered vulnerability remediation

**Secret Protection**:
- **Secret Scanning**: Detect exposed credentials and tokens
- **Push Protection**: Prevent secret commits
- **Non-provider Pattern Detection**: Find custom secret formats

## Community Health Files

### Default Community Health Files
Create a public `.github` repository with standardized files that apply across all repositories:

**Essential Files**:
- `CODE_OF_CONDUCT.md` - Community behavior standards
- `CONTRIBUTING.md` - Contribution guidelines and processes
- `SECURITY.md` - Vulnerability reporting instructions
- `SUPPORT.md` - Support resources and contact information
- Issue and pull request templates
- Discussion category forms
- Funding/sponsor information

**Implementation**:
- Files in `.github` repository apply to all repositories lacking their own versions
- Priority order: `.github` folder → repository root → `docs` folder
- Repository-specific files override defaults
- Must be in public repository to work organization-wide

### License and Legal Considerations
- Add appropriate open source license
- License files cannot be defaulted (must be per-repository)
- Clear intellectual property guidelines
- Contribution licensing terms

## Implementation Roadmap

### Phase 1: Foundation (Priority: High)
1. **Set Repository Access Controls**
   - Configure visibility settings
   - Establish permission structure
   - Enable branch protection

2. **Create Security Policy**
   - Add `SECURITY.md` with vulnerability reporting process
   - Include contact information and response expectations
   - Link from README and other documentation

3. **Enable Basic Security Features**
   - Turn on dependency graph
   - Enable Dependabot alerts
   - Configure Dependabot security updates

### Phase 2: Advanced Security (Priority: Medium)
1. **Implement Code Security**
   - Enable CodeQL code scanning (default setup)
   - Configure secret scanning and push protection
   - Set up dependency review for pull requests

2. **Establish Community Health**
   - Create `.github` repository with default community files
   - Add code of conduct and contributing guidelines
   - Set up issue and pull request templates

3. **Configure Automated Updates**
   - Enable Dependabot version updates
   - Create custom `dependabot.yml` configuration
   - Set up auto-triage rules for alert management

### Phase 3: Advanced Processes (Priority: Lower)
1. **Vulnerability Management Workflow**
   - Train team on security advisory creation
   - Establish CVD process with security researchers
   - Create incident response procedures

2. **Community Engagement**
   - Add support resources and documentation
   - Create discussion categories and forms
   - Implement contributor recognition systems

## Key Considerations

### Security Best Practices
- **Defense in Depth**: Layer multiple security controls
- **Principle of Least Privilege**: Minimal necessary permissions
- **Regular Updates**: Keep dependencies and tools current
- **Monitoring**: Continuous security assessment and alerting
- **Documentation**: Clear, accessible security policies

### Community Management
- **Transparency**: Open communication about security practices
- **Responsiveness**: Timely handling of security reports
- **Recognition**: Credit contributors and security researchers
- **Education**: Help community understand security practices

### Compliance Considerations
- CVE assignment and tracking
- Security advisory publication
- Audit trails for security incidents
- Integration with security toolchains

## Monitoring and Maintenance

### Regular Activities
- Review and respond to Dependabot alerts
- Monitor code scanning results
- Update security policies as needed
- Audit repository access permissions
- Review and update community health files

### Metrics to Track
- Time to vulnerability disclosure
- Number of security alerts resolved
- Community engagement with security policies
- Coverage of security scanning tools
- Response time to security reports

# Resources
- [GitHub: Adding a security policy to your repository](https://docs.github.com/en/code-security/getting-started/adding-a-security-policy-to-your-repository)
- [GitHub: About coordinated disclosure of security vulnerabilities](https://docs.github.com/en/code-security/security-advisories/guidance-on-reporting-and-writing-information-about-vulnerabilities/about-coordinated-disclosure-of-security-vulnerabilities#about-reporting-and-disclosing-vulnerabilities-in-projects-on-github)
- [GitHub: About repository security advisories](https://docs.github.com/en/code-security/security-advisories/working-with-repository-security-advisories/about-repository-security-advisories)
- [GitHub: Quickstart for securing your repository](https://docs.github.com/en/code-security/getting-started/quickstart-for-securing-your-repository)
- [GitHub: Setting up your project for healthy contributions](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions)
- [GitHub: Creating a default community health file](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/creating-a-default-community-health-file)

**Comment by DrAcula27** [comment]
# Comments - Security & Protections
1. It looks like the gist of the security policy is just having what GitHub calls _'code health'_ documents like `SECURITY.md`, `CONTRIBUTING.md`, and `CODE_OF_CONDUCT.md`.
1. I saw that Tracy started creating a security policy in the repo, but I do not have permissions to edit it directly.
    1. GitHub actually forces you to fork the repo for this.
    1. Since I already work with a forked version of this repo, I started on a security branch for setting up a default `SECURITY.md` file.
1. I don't have permissions to set up branch protections. I think @ryanfchase or @cottonchristopher will have to do that.

**Comment by DrAcula27** [comment]
@ryanfchase do you want to go ahead and update the `dev` branch then?

**Comment by ryanfchase** [comment]
@DrAcula27 also I think we should move forward with a new ticket(s) to implement the changes you've outlined in your report. Thanks for writing that up!

> ## Implementation Roadmap
> 
> ### Phase 1: Foundation (Priority: High)
> 1. **Set Repository Access Controls**
>    - Configure visibility settings
>    - Establish permission structure
>    - Enable branch protection
> 
> 2. **Create Security Policy**
>    - Add `SECURITY.md` with vulnerability reporting process
>    - Include contact information and response expectations
>    - Link from README and other documentation
> 
> 3. **Enable Basic Security Features**
>    - Turn on dependency graph
>    - Enable Dependabot alerts
>    - Configure Dependabot security updates
> 
> ### Phase 2: Advanced Security (Priority: Medium)
> 1. **Implement Code Security**
>    - Enable CodeQL code scanning (default setup)
>    - Configure secret scanning and push protection
>    - Set up dependency review for pull requests
> 
> 2. **Establish Community Health**
>    - Create `.github` repository with default community files
>    - Add code of conduct and contributing guidelines
>    - Set up issue and pull request templates
> 
> 3. **Configure Automated Updates**
>    - Enable Dependabot version updates
>    - Create custom `dependabot.yml` configuration
>    - Set up auto-triage rules for alert management
> 
> ### Phase 3: Advanced Processes (Priority: Lower)
> 1. **Vulnerability Management Workflow**
>    - Train team on security advisory creation
>    - Establish CVD process with security researchers
>    - Create incident response procedures
> 
> 2. **Community Engagement**
>    - Add support resources and documentation
>    - Create discussion categories and forms
>    - Implement contributor recognition systems

**Comment by ryanfchase** [comment]
I created this ticket to handle SECURITY.md:
- https://github.com/hackforla/311-data/issues/2032

**Comment by ryanfchase** [comment]
@DrAcula27 thank you for your work on this ticket! with the [above tickets created](https://github.com/hackforla/311-data/issues/1867#issuecomment-3287781686), we can close this ticket 😄
